src/Stack/Build/ConstructPlan.hs

```diff
@@ -192,6 +188,16 @@ constructPlan mbp0 baseConfigOpts0 locals extraToBuild0 localDumpPkgs loadPackag
     -- or local packages.
     toolMap = getToolMap mbp0
 
+withLoadPackageVersions
+    :: (MonadIO m, HasHttpManager env, MonadReader env m, MonadBaseControl IO m, MonadCatch m, MonadLogger m, HasConfig env)
+    => ((PackageName -> IO (Set Version)) -> m a)
+    -> m a
+withLoadPackageVersions inner =
+    join $ runOnce $ do
```

Comment:
Oh, I realized that the `runOnce` wouldn't help and removed it in my second commit, but I hadn't thought that it was quite so bad!

I'll just squash my two commits and hope that the combination makes more sense.